cloudinary.go

```diff
@@ -0,0 +1,390 @@
+func (r *Cloudinary) Put(file, content string) error {
+	tempFile, err := r.tempFile(content)
+	defer os.Remove(tempFile.Name())
+	if err != nil {
+		return err
+	}
+	_, err = r.instance.Upload.Upload(r.ctx, tempFile.Name(), uploader.UploadParams{
+		PublicID:       r.getPublicId(file),
+		UseFilename:    api.Bool(true),
+		UniqueFilename: api.Bool(false),
+		ResourceType:   r.getResourceType(file),
+	})
+	return err
+}
+
+// PutFile stores a new file on the disk.
+func (r *Cloudinary) PutFile(path string, source filesystem.File) (string, error) {
+	return r.PutFileAs(path, source, source.File())
+}
+
+// PutFileAs stores a new file on the disk.
+func (r *Cloudinary) PutFileAs(path string, source filesystem.File, name string) (string, error) {
+	uploadResult, err := r.instance.Upload.Upload(r.ctx, source.File(), uploader.UploadParams{
+		Folder:         validPath(path),
+		PublicID:       r.getPublicId(name),
+		UseFilename:    api.Bool(true),
+		UniqueFilename: api.Bool(false),
+	})
+	if err != nil {
+		return "", err
+	}
+	return uploadResult.PublicID, nil
+}
+
+// Size returns the file size of a given file.
+func (r *Cloudinary) Size(file string) (int64, error) {
+	resource, err := r.getAsset(file)
+	if err != nil {
+		return 0, err
+	}
+	return int64(resource.Bytes), nil
+}
+
+// TemporaryUrl get the temporary url of a file.
+func (r *Cloudinary) TemporaryUrl(file string, time time.Time) (string, error) {
+	return "", errors.New("we don't support temporary url for cloudinary")
+}
+
+// WithContext sets the context for the driver.
+func (r *Cloudinary) WithContext(ctx context.Context) filesystem.Driver {
+	driver, err := NewCloudinary(ctx, r.config, r.disk)
+	if err != nil {
+		color.Redf("[Cloudinary] init disk error: %+v\n", err)
+		return nil
+	}
+	return driver
+}
+
+// Url returns the url for a file.
+func (r *Cloudinary) Url(file string) string {
+	asset, err := r.getAsset(file)
+	if err != nil {
+		return ""
+	}
+	return asset.SecureURL
+}
+
+func (r *Cloudinary) getAsset(path string) (*admin.AssetResult, error) {
+	return r.instance.Admin.Asset(r.ctx, admin.AssetParams{
+		PublicID:  r.getPublicId(path),
+		AssetType: api.AssetType(r.getResourceType(path)),
+	})
+}
+
+func (r *Cloudinary) tempFile(content string) (*os.File, error) {
+	tempFile, err := os.CreateTemp(os.TempDir(), "goravel-")
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := tempFile.WriteString(content); err != nil {
+		return nil, err
+	}
+
+	return tempFile, nil
+}
+
+func (r *Cloudinary) getResourceType(path string) string {
+	extension := strings.TrimPrefix(filepath.Ext(path), ".")
+	value := "image"
+	resourceTypes := r.config.Get(fmt.Sprintf("filesystems.disks.%s.resource_types", r.disk)).(map[string][]string)
+
+	for resourceType, extensions := range resourceTypes {
+		for _, ext := range extensions {
+			if ext == extension {
+				value = resourceType
+				break
+			}
+		}
+	}
+
+	return value
+}
+
+func (r *Cloudinary) getPublicId(path string) string {
```

Comment:
Q: Can't PublicId contain extension?

Reply:
No, PublicId doesn't contain extension other than `raw` type.

Reply:
What's the PublicId looks like?

Reply:
> - The Public ID value for images and videos should not include a file extension. Include the file extension for raw files only.
> - Can be up to 255 characters, including non-English characters, periods (.), forward slashes (/), underscores (_), hyphens (-).
> -  Public ID values cannot begin or end with a space or forward slash (/). Additionally, they cannot include the following characters: ? & # \ % < > +

ref: https://support.cloudinary.com/hc/en-us/community/posts/360000303551/comments/7949396230802

Reply:
Okay, thanks!